app/src/org/commcare/dalvik/odk/provider/InstanceProvider.java

```diff
@@ -97,13 +111,13 @@ public boolean onCreate() {
         return true;
     }
 
-    /**
-     * Setup helper to access database.
-     */
     void init() {
-        //this is terrible, we need to be binding to the cc service, etc. Temporary code for testing
-        if(mDbHelper == null) {
-            mDbHelper = new DatabaseHelper(CommCareApplication._(), DATABASE_NAME);
+        // this is terrible, we need to be binding to the cc service, etc. Temporary code for testing
+        if (mDbHelper == null || mDbHelper.getAppId() != ProviderUtils.getSeatedOrInstallingAppId()) {
+            String dbName = ProviderUtils.getProviderDbName(
+                    ProviderUtils.ProviderType.INSTANCES,
+                    ProviderUtils.getSeatedOrInstallingAppId());
+            mDbHelper = new DatabaseHelper(CommCareApplication._(), dbName);
```

Comment:
I'm a bit confused. Why don't we pass this the DB name + the app_id and have it seat against that, rather than reading the app ID from the DB name?